File: binding/src/network/libqtcpserver.cpp
#include <QHostAddress>
#include <QNetworkProxy>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QTcpServer>
#include <QTcpSocket>
#include <qtcpserver.h>
#include "libqtcpserver.h"
#include "libqtcpserver.hxx"

QTcpServer* QTcpServer_new() {
	 return new VirtualQTcpServer();
}

QTcpServer* QTcpServer_new2(QObject* parent) {
	 return new VirtualQTcpServer(parent);
}

libqt_string QTcpServer_Tr(const char* s) {
	QString _ret = QTcpServer::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QTcpServer_Listen(QTcpServer* self) {
	return self->listen();
}

void QTcpServer_Close(QTcpServer* self) {
	self->close();
}

bool QTcpServer_IsListening(const QTcpServer* self) {
	return self->isListening();
}

void QTcpServer_SetMaxPendingConnections(QTcpServer* self, int numConnections) {
	self->setMaxPendingConnections(numConnections);
}

int QTcpServer_MaxPendingConnections(const QTcpServer* self) {
	return self->maxPendingConnections();
}

void QTcpServer_SetListenBacklogSize(QTcpServer* self, int size) {
	self->setListenBacklogSize(size);
}

int QTcpServer_ListenBacklogSize(const QTcpServer* self) {
	return self->listenBacklogSize();
}

uint16_t QTcpServer_ServerPort(const QTcpServer* self) {
	return self->serverPort();
}

QHostAddress* QTcpServer_ServerAddress(const QTcpServer* self) {
	return new QHostAddress(self->serverAddress());
}

intptr_t QTcpServer_SocketDescriptor(const QTcpServer* self) {
	return self->socketDescriptor();
}

bool QTcpServer_SetSocketDescriptor(QTcpServer* self, intptr_t socketDescriptor) {
	return self->setSocketDescriptor(socketDescriptor);
}

bool QTcpServer_WaitForNewConnection(QTcpServer* self) {
	return self->waitForNewConnection();
}

bool QTcpServer_HasPendingConnections(const QTcpServer* self) {
	return self->hasPendingConnections();
}

QTcpSocket* QTcpServer_NextPendingConnection(QTcpServer* self) {
	return self->nextPendingConnection();
}

int QTcpServer_ServerError(const QTcpServer* self) {
	return self->serverError();
}

libqt_string QTcpServer_ErrorString(const QTcpServer* self) {
	QString _ret = self->errorString();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QTcpServer_PauseAccepting(QTcpServer* self) {
	self->pauseAccepting();
}

void QTcpServer_ResumeAccepting(QTcpServer* self) {
	self->resumeAccepting();
}

void QTcpServer_SetProxy(QTcpServer* self, const QNetworkProxy* networkProxy) {
	self->setProxy(*networkProxy);
}

QNetworkProxy* QTcpServer_Proxy(const QTcpServer* self) {
	return new QNetworkProxy(self->proxy());
}

void QTcpServer_NewConnection(QTcpServer* self) {
	self->newConnection();
}

void QTcpServer_Connect_NewConnection(QTcpServer* self, intptr_t slot) {
    void (*slotFunc)(QTcpServer*) = reinterpret_cast<void (*)(QTcpServer*)>(slot);
    QTcpServer::connect(self, &QTcpServer::newConnection, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QTcpServer_AcceptError(QTcpServer* self, int socketError) {
	self->acceptError(static_cast<QAbstractSocket::SocketError>(socketError));
}

void QTcpServer_Connect_AcceptError(QTcpServer* self, intptr_t slot) {
    void (*slotFunc)(QTcpServer*, int) = reinterpret_cast<void (*)(QTcpServer*, int)>(slot);
    QTcpServer::connect(self, &QTcpServer::acceptError, [self, slotFunc](QAbstractSocket::SocketError socketError) {
	slotFunc(self, socketError);
    });
}

libqt_string QTcpServer_Tr2(const char* s, const char* c) {
	QString _ret = QTcpServer::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QTcpServer_Tr3(const char* s, const char* c, int n) {
	QString _ret = QTcpServer::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

bool QTcpServer_Listen1(QTcpServer* self, const QHostAddress* address) {
	return self->listen(*address);
}

bool QTcpServer_Listen2(QTcpServer* self, const QHostAddress* address, uint16_t port) {
	return self->listen(*address, port);
}

bool QTcpServer_WaitForNewConnection1(QTcpServer* self, int msec) {
	return self->waitForNewConnection(msec);
}

bool QTcpServer_WaitForNewConnection2(QTcpServer* self, int msec, bool* timedOut) {
	return self->waitForNewConnection(msec, timedOut);
}

// Base class handler implementation
bool QTcpServer_QBaseHasPendingConnections(const QTcpServer* self) {
	auto* vqtcpserver = dynamic_cast<const VirtualQTcpServer*>(self);
	if (vqtcpserver && vqtcpserver->isVirtualQTcpServer) {
vqtcpserver->setQTcpServer_HasPendingConnections_IsBase(true);
	return vqtcpserver->hasPendingConnections();
}
}

// Auxiliary method to allow providing re-implementation
void QTcpServer_OnHasPendingConnections(const QTcpServer* self, intptr_t slot) {
	auto* vqtcpserver = dynamic_cast<const VirtualQTcpServer*>(self);
	if (vqtcpserver && vqtcpserver->isVirtualQTcpServer) {
vqtcpserver->setQTcpServer_HasPendingConnections_Callback(reinterpret_cast<VirtualQTcpServer::QTcpServer_HasPendingConnections_Callback>(slot));
}
}

// Base class handler implementation
QTcpSocket* QTcpServer_QBaseNextPendingConnection(QTcpServer* self) {
	auto* vqtcpserver = dynamic_cast<VirtualQTcpServer*>(self);
	if (vqtcpserver && vqtcpserver->isVirtualQTcpServer) {
vqtcpserver->setQTcpServer_NextPendingConnection_IsBase(true);
	return vqtcpserver->nextPendingConnection();
}
}

// Auxiliary method to allow providing re-implementation
void QTcpServer_OnNextPendingConnection(QTcpServer* self, intptr_t slot) {
	auto* vqtcpserver = dynamic_cast<VirtualQTcpServer*>(self);
	if (vqtcpserver && vqtcpserver->isVirtualQTcpServer) {
vqtcpserver->setQTcpServer_NextPendingConnection_Callback(reinterpret_cast<VirtualQTcpServer::QTcpServer_NextPendingConnection_Callback>(slot));
}
}

void QTcpServer_Connect_pendingConnectionAvailable(QTcpServer* self, intptr_t slot) {
    void (*slotFunc)(QTcpServer*) = reinterpret_cast<void (*)(QTcpServer*)>(slot);
    QTcpServer::connect(self, &QTcpServer::pendingConnectionAvailable, [self, slotFunc]() {
	slotFunc(self);
    });
}

void QTcpServer_Delete(QTcpServer* self) {
    delete self;
}

File: binding/include/network/libqtcpserver.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qtcpserver.html
pub const qtcpserver = struct {

    /// New constructs a new QTcpServer object.
    pub fn New() QtC.QTcpServer {
        return qtc.QTcpServer_new();
    }


    /// New2 constructs a new QTcpServer object.
    pub fn New2(parent: ?*anyopaque) QtC.QTcpServer {
        return qtc.QTcpServer_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QTcpServer_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtcpserver.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Listen(self: ?*anyopaque, ) bool {
        return qtc.QTcpServer_Listen(@ptrCast(self));
    }

    pub fn Close(self: ?*anyopaque, ) void {
        qtc.QTcpServer_Close(@ptrCast(self));
    }

    pub fn IsListening(self: ?*anyopaque, ) bool {
        return qtc.QTcpServer_IsListening(@ptrCast(self));
    }

    pub fn SetMaxPendingConnections(self: ?*anyopaque, numConnections: i32) void {
        qtc.QTcpServer_SetMaxPendingConnections(@ptrCast(self), numConnections);
    }

    pub fn MaxPendingConnections(self: ?*anyopaque, ) i32 {
        return qtc.QTcpServer_MaxPendingConnections(@ptrCast(self));
    }

    pub fn SetListenBacklogSize(self: ?*anyopaque, size: i32) void {
        qtc.QTcpServer_SetListenBacklogSize(@ptrCast(self), size);
    }

    pub fn ListenBacklogSize(self: ?*anyopaque, ) i32 {
        return qtc.QTcpServer_ListenBacklogSize(@ptrCast(self));
    }

    pub fn ServerPort(self: ?*anyopaque, ) u16 {
        return qtc.QTcpServer_ServerPort(@ptrCast(self));
    }

    pub fn ServerAddress(self: ?*anyopaque, ) QtC.QHostAddress {
        return qtc.QTcpServer_ServerAddress(@ptrCast(self));
    }

    pub fn SocketDescriptor(self: ?*anyopaque, ) isize {
        return qtc.QTcpServer_SocketDescriptor(@ptrCast(self));
    }

    pub fn SetSocketDescriptor(self: ?*anyopaque, socketDescriptor: isize) bool {
        return qtc.QTcpServer_SetSocketDescriptor(@ptrCast(self), socketDescriptor);
    }

    pub fn WaitForNewConnection(self: ?*anyopaque, ) bool {
        return qtc.QTcpServer_WaitForNewConnection(@ptrCast(self));
    }

    pub fn HasPendingConnections(self: ?*anyopaque, ) bool {
        return qtc.QTcpServer_HasPendingConnections(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnHasPendingConnections(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) bool) void {
        qtc.QTcpServer_OnHasPendingConnections(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseHasPendingConnections(self: ?*anyopaque, ) bool {
        return qtc.QTcpServer_QBaseHasPendingConnections(@ptrCast(self));
    }

    pub fn NextPendingConnection(self: ?*anyopaque, ) QtC.QTcpSocket {
        return qtc.QTcpServer_NextPendingConnection(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnNextPendingConnection(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) QtC.QTcpSocket) void {
        qtc.QTcpServer_OnNextPendingConnection(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseNextPendingConnection(self: ?*anyopaque, ) QtC.QTcpSocket {
        return qtc.QTcpServer_QBaseNextPendingConnection(@ptrCast(self));
    }

    pub fn ServerError(self: ?*anyopaque, ) i32 {
        return qtc.QTcpServer_ServerError(@ptrCast(self));
    }

    pub fn ErrorString(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QTcpServer_ErrorString(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtcpserver.ErrorString: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn PauseAccepting(self: ?*anyopaque, ) void {
        qtc.QTcpServer_PauseAccepting(@ptrCast(self));
    }

    pub fn ResumeAccepting(self: ?*anyopaque, ) void {
        qtc.QTcpServer_ResumeAccepting(@ptrCast(self));
    }

    pub fn SetProxy(self: ?*anyopaque, networkProxy: ?*anyopaque) void {
        qtc.QTcpServer_SetProxy(@ptrCast(self), @ptrCast(networkProxy));
    }

    pub fn Proxy(self: ?*anyopaque, ) QtC.QNetworkProxy {
        return qtc.QTcpServer_Proxy(@ptrCast(self));
    }

    pub fn IncomingConnection(self: ?*anyopaque, handle: isize) void {
        qtc.QTcpServer_IncomingConnection(@ptrCast(self), handle);
    }

    /// Allows for overriding the related default method
    pub fn OnIncomingConnection(self: ?*anyopaque, callback: *const fn (?*anyopaque, isize) callconv(.c) void) void {
        qtc.QTcpServer_OnIncomingConnection(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseIncomingConnection(self: ?*anyopaque, handle: isize) void {
        qtc.QTcpServer_QBaseIncomingConnection(@ptrCast(self), handle);
    }

    pub fn AddPendingConnection(self: ?*anyopaque, socket: ?*anyopaque) void {
        qtc.QTcpServer_AddPendingConnection(@ptrCast(self), @ptrCast(socket));
    }

    /// Allows for overriding the related default method
    pub fn OnAddPendingConnection(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QTcpServer_OnAddPendingConnection(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseAddPendingConnection(self: ?*anyopaque, socket: ?*anyopaque) void {
        qtc.QTcpServer_QBaseAddPendingConnection(@ptrCast(self), @ptrCast(socket));
    }

    pub fn NewConnection(self: ?*anyopaque, ) void {
        qtc.QTcpServer_NewConnection(@ptrCast(self));
    }

    pub fn OnNewConnection(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QTcpServer_Connect_NewConnection(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn AcceptError(self: ?*anyopaque, socketError: i32) void {
        qtc.QTcpServer_AcceptError(@ptrCast(self), @intCast(socketError));
    }

    pub fn OnAcceptError(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QTcpServer_Connect_AcceptError(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QTcpServer_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtcpserver.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QTcpServer_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qtcpserver.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Listen1(self: ?*anyopaque, address: ?*anyopaque) bool {
        return qtc.QTcpServer_Listen1(@ptrCast(self), @ptrCast(address));
    }

    pub fn Listen2(self: ?*anyopaque, address: ?*anyopaque, port: u16) bool {
        return qtc.QTcpServer_Listen2(@ptrCast(self), @ptrCast(address), port);
    }

    pub fn WaitForNewConnection1(self: ?*anyopaque, msec: i32) bool {
        return qtc.QTcpServer_WaitForNewConnection1(@ptrCast(self), msec);
    }

    pub fn WaitForNewConnection2(self: ?*anyopaque, msec: i32, timedOut: *bool) bool {
        return qtc.QTcpServer_WaitForNewConnection2(@ptrCast(self), msec, @ptrCast(timedOut));
    }

    /// Wrapper to allow calling private signal
    pub fn OnpendingConnectionAvailable(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) void) void {
        qtc.QTcpServer_Connect_pendingConnectionAvailable(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QTcpServer_Delete(@ptrCast(self));
    }
};
